sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -426,3 +426,68 @@ case class MinOf(left: Expression, right: Expression) extends BinaryArithmetic {
 
   override def toString: String = s"MinOf($left, $right)"
 }
+
+case class Pmod(left: Expression, right: Expression) extends BinaryArithmetic {
+
+  override def symbol: String = "pmod"
+
+  override def toString: String = s"pmod($left, $right)"
+
+  protected def checkTypesInternal(t: DataType) =
+    TypeUtils.checkForNumericExpr(t, "pmod")
+
+  override lazy val resolved =
+    childrenResolved && checkInputDataTypes().isSuccess && !DecimalType.isFixed(dataType)
+
+  protected override def evalInternal(left: Any, right: Any) =
+    dataType match {
+      case IntegerType => pmod(left.asInstanceOf[Integer], right.asInstanceOf[Integer])
+      case LongType => pmod(left.asInstanceOf[Long], right.asInstanceOf[Long])
+      case ShortType => pmod(left.asInstanceOf[Short], right.asInstanceOf[Short])
+      case ByteType => pmod(left.asInstanceOf[Byte], right.asInstanceOf[Byte])
+      case FloatType => pmod(left.asInstanceOf[Float], right.asInstanceOf[Float])
+      case DoubleType => pmod(left.asInstanceOf[Double], right.asInstanceOf[Double])
+      case _: DecimalType => pmod(left.asInstanceOf[Decimal], right.asInstanceOf[Decimal])
+    }
+
+  val decimalAddMethod = "$plus"
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = dataType match {
+    case dt: DecimalType =>
+      defineCodeGen(ctx, ev, (eval1, eval2) =>
+        s"""($eval1.remainder($eval2).$decimalAddMethod($eval2)).remainder($eval2)""")
+    case _ =>
+      defineCodeGen(ctx, ev, (eval1, eval2) => s"""(((${ctx.javaType(dataType)})$eval1
+                                     % ((${ctx.javaType(dataType)})$eval2)
+                                     + ((${ctx.javaType(dataType)})$eval2))
+                                     % ((${ctx.javaType(dataType)})$eval2)) """)
+  }
```

Comment:
Do we really need so many type casts? For byte and short, we only need to cast the final result:`(byte)((a % n + n) % n)`, For int, long, float, double, we can just write `(a % n + n) % n`.